docs/source/working_with_aiida/caching.rst

```diff
@@ -65,88 +115,76 @@ This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str
-
-Finally, workchains can create links not only to nodes which they create themselves, but also to nodes created by a calculation that they called, or even their ancestors. This is where caching becomes impossible. Consider the following example (using workfunctions for simplicity):
-
-.. code:: python
+.. _caching_limitations:
 
-    from aiida.engine import workfunction
-    from aiida.orm import Int
+Limitations
+-----------
 
-    @workfunction
-    def select(a, b):
-        return b
+#. While caching saves unnecessary computations, the current implementation does not yet save disk space:
+   The output nodes of the cached calculation are full copies of the original outputs.
+   The plan is to add data deduplication as a global feature at the repository and database level (independent of caching).
 
-    d = Int(1)
-    r1 = select(d, d)
-    r2 = select(Int(1), Int(1))
+#. Workflow nodes are not cached. As explained below, in the current design this follows from the requirement that the provenance graph be independent of whether caching is enabled or not.
 
-The two ``select`` workfunctions have the same inputs as far as their hashes go. However, the first call uses the same input node twice, while the second one has two different inputs. If the second call should be cached from the first one, it is not clear which of the two input nodes should be returned.
+   * **Calculation nodes:** Calculation nodes can have data inputs and create new data nodes as outputs.
+     In order to make it look as if a cloned calculation produced its own outputs, the output nodes are copied and linked as well.
+   * **Workflow nodes:** Workflows differ from calculations in that they can *return* an input node or an output node created by a calculation.
+     Since caching does not care about the *identity* of input nodes but only their *content*, it is not straightforward to figure out which node to return in a cached workflow.
 
-While this example might seem contrived, the conclusion is valid more generally: Because workchains can return nodes from their history, they cannot be cached. Since even two equivalent workchains (with the same inputs) can have a different history, there is no way to deduce which links should be created on the new workchain without actually running it.
+  For the moment, this limitation is acceptable since the runtime of AiiDA WorkChains is usually dominated by expensive calculations, which are covered by the current caching mechanism.
 
-Overall, this limitation is acceptable: The runtime of AiiDA workchains is usually dominated by time spent inside expensive calculations. Since these can be avoided with the caching mechanism, it still improves the runtime and required computer resources a lot.
+#. The caching mechanism for calculations *should* trigger only when the inputs and the calculation to be performed are exactly the same.
```

Comment:
I think we discussed this before somewhere but cannot find it. I think the problem is worse as in that only the parser name is considered through the `parser_name` attribute. That is to say that as long as that stays the same, the parser can be changed completely and the hash wouldn't change.

Reply:
Yes, I've seen your comment on this - I'll resolve this in order not to duplicate things